python/cudf/cudf/core/join/_join_helpers.py

```diff
@@ -0,0 +1,194 @@
+                obj._index._data[self.name] = value
+            else:
+                raise KeyError()
+
+
+def _frame_select_by_indexers(
+    frame: Frame, indexers: Iterable[_Indexer]
+) -> Frame:
+    # Select columns from the given `Frame` using `indexers`,
+    # and return a new `Frame`.
+    index_data = frame._data.__class__()
+    data = frame._data.__class__()
+
+    for idx in indexers:
+        if idx.index:
+            index_data[idx.name] = idx.get(frame)
+        else:
+            data[idx.name] = idx.get(frame)
+
+    result_index = cudf.Index._from_data(index_data) if index_data else None
+    result = cudf.core.frame.Frame(data=data, index=result_index)
+    return result
+
+
+def _match_join_keys(lcol: ColumnBase, rcol: ColumnBase, how: str) -> Dtype:
+    # cast the keys lcol and rcol to a common dtype
+
+    ltype = lcol.dtype
+    rtype = rcol.dtype
+
+    # if either side is categorical, different logic
+    if isinstance(ltype, CategoricalDtype) or isinstance(
+        rtype, CategoricalDtype
+    ):
+        return _match_categorical_dtypes(ltype, rtype, how)
+
+    if pd.api.types.is_dtype_equal(ltype, rtype):
+        return ltype
+
+    if (np.issubdtype(ltype, np.number)) and (np.issubdtype(rtype, np.number)):
+        common_type = (
+            max(ltype, rtype)
+            if ltype.kind == rtype.kind
+            else np.find_common_type([], (ltype, rtype))
+        )
+
+    elif np.issubdtype(ltype, np.datetime64) and np.issubdtype(
+        rtype, np.datetime64
+    ):
+        common_type = max(ltype, rtype)
+
+    if how == "left":
+        if rcol.fillna(0).can_cast_safely(ltype):
+            return ltype
+        else:
+            warnings.warn(
+                f"Can't safely cast column from {rtype} to {ltype}, "
+                "upcasting to {common_type}."
+            )
+
+    if common_type:
+        return common_type
+
+    return None
+
+
+def _match_categorical_dtypes(ltype: Dtype, rtype: Dtype, how: str) -> Dtype:
+    # cast the keys lcol and rcol to a common dtype
+    # when at least one of them is a categorical type
+
+    if isinstance(ltype, CategoricalDtype) and isinstance(
+        rtype, CategoricalDtype
+    ):
+        # if both are categoricals, logic is complicated:
+        return _match_categorical_dtypes_both(ltype, rtype, how)
+
+    if isinstance(ltype, CategoricalDtype):
+        if how in {"left", "leftsemi", "leftanti"}:
+            return ltype
+        common_type = ltype.categories.dtype
+    elif isinstance(rtype, CategoricalDtype):
+        common_type = rtype.categories.dtype
+    return common_type
+
+
+def _match_categorical_dtypes_both(
+    ltype: CategoricalDtype, rtype: CategoricalDtype, how: str
+) -> Dtype:
+    # The commontype depends on both `how` and the specifics of the
+    # categorical variables to be merged.
+
+    # when both are ordered and both have the same categories,
+    # no casting required:
+    if ltype == rtype:
```

Comment:
Conundrum here. I _think_ this is redundant with the check at the start of `_match_join_keys`, which should catch this case. The extra check could be expensive since it's a binop over the `categories`.  That said, this function _could_ be called outside of that context, so it's not technically correct unless it has the check. Is there some better way of avoiding checking twice?

Reply:
This sounds like a reason to keep the ordering mentioned [here](https://github.com/rapidsai/cudf/pull/7454#discussion_r589824295)?

Reply:
So if we have to categorical dtypes that are exactly the same, we'll fall into `_match_categorical_dtypes`, it will check once, and return `ltype`, thereby avoiding this second check. But if they're not categorical, it won't even enter `_match_categorical_dtypes`, it *wont* do the check inside that function, but will return `ltype` here, right?

Reply:
right - so we only ever do the equality check once